### PR TITLE
mgr/cephadm: retry mgr fail over in case of transient failure

### PR DESCRIPTION
**Pull request by adk3798**
Fixes: https://tracker.ceph.com/issues/55279

Signed-off-by: Adam King <adking@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by adk3798** [comment]
http://pulpito.ceph.com/adking-2022-04-13_05:16:48-orch:cephadm-wip-adk-testing-2022-04-12-2231-distro-basic-smithi/

2 Failures

- https://tracker.ceph.com/issues/44587
- ceph-volume lvm batch failure `ERROR: error creating empty object store in /var/lib/ceph/osd/ceph-3/: (5) Input/output error` that may be a faulty hardware thing, we're not sure. Did a rerun of the osds tests https://pulpito.ceph.com/adking-2022-04-13_13:12:12-orch:cephadm:osds-wip-adk-testing-2022-04-12-2231-distro-basic-smithi/ and the test that failed in the last run passed and the one test that did fail is due to https://tracker.ceph.com/issues/49287

To summarize, run was passing. Will start tracking this type of lvm batch failure to see if it happens on the same hosts (which would imply a possible hardware issue) in https://tracker.ceph.com/issues/55319

**Review by rkachach — APPROVED** [review]
As I understand this change is to give some resilience against transient errors but it would be good also to investigate what causes these errors. Just to make sure we understand the root cause of these errors.

File: src/pybind/mgr/cephadm/services/cephadmservice.py [hunk 1]
```diff
@@ -672,19 +673,32 @@ def get_active_daemon(self, daemon_descrs: List[DaemonDescription]) -> DaemonDes
...
+                # fail over
+                ret, out, err = self.mgr.check_mon_command({
+                    'prefix': 'mgr fail',
+                    'who': self.mgr.get_mgr_id(),
+                })
+                return
+            except OrchestratorError as e:
+                if 'Need standby mgr daemon' not in str(e):
+                    raise e
+                time.sleep(sleep_secs)
```
Comment (mgfritch) [inline src/pybind/mgr/cephadm/services/cephadmservice.py:700]:
```suggestion
            except NoStandbyError as e:
                if 'Need standby mgr daemon' not in str(e):
                    raise e
                time.sleep(sleep_secs)
```
subclass this?

File: src/pybind/mgr/cephadm/services/cephadmservice.py [hunk 2]
```diff
@@ -672,19 +673,32 @@ def get_active_daemon(self, daemon_descrs: List[DaemonDescription]) -> DaemonDes
...
-            'prefix': 'mgr fail',
-            'who': self.mgr.get_mgr_id(),
-        })
+        # this has been seen to sometimes transiently fail even when there are multiple
+        # mgr daemons. As long as there are multiple known mgr daemons, we should retry.
+        no_standby_exc = OrchestratorError('Need standby mgr daemon', event_kind_subject=(
+            'daemon', 'mgr' + self.mgr.get_mgr_id()))
+        if len([d for d in self.mgr.cache.get_daemons_by_type('mgr')]) < 2:
+            # only one known mgr daemon, retry won't help us
+            raise no_standby_exc
```
Comment (mgfritch) [inline src/pybind/mgr/cephadm/services/cephadmservice.py:682]: seems like this could also cause transient upgrade failures, maybe the check during `mgr_map_has_standby` is sufficient?

Comment (adk3798) [reply src/pybind/mgr/cephadm/services/cephadmservice.py:682]: my thinking was, since this was appearing during upgrade and we already have a hard check for 2 mgr daemons marked running in the cache there then it must not be what we have cached that's the issue. However, thinking about it, those checks existing also makes this redundant anyway so I should really just remove it anyhow.

File: src/pybind/mgr/cephadm/services/cephadmservice.py [hunk 3]
```diff
@@ -672,19 +673,29 @@ def get_active_daemon(self, daemon_descrs: List[DaemonDescription]) -> DaemonDes
...
-            'prefix': 'mgr fail',
-            'who': self.mgr.get_mgr_id(),
-        })
+        # this has been seen to sometimes transiently fail even when there are multiple
+        # mgr daemons. As long as there are multiple known mgr daemons, we should retry.
+        class NoStandbyError(OrchestratorError):
+            pass
+        no_standby_exc = NoStandbyError('Need standby mgr daemon', event_kind_subject=(
+            'daemon', 'mgr' + self.mgr.get_mgr_id()))
+        for sleep_secs in [1, 3, 6]:
```
Comment (mgfritch) [inline src/pybind/mgr/cephadm/services/cephadmservice.py:682]: (nit) wondering if we will need to wait for a longer duration than this?

**Review by mgfritch — APPROVED** [review]

File: src/pybind/mgr/cephadm/services/cephadmservice.py (hunk 3 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/services/cephadmservice.py]: Upped it to 2, 8, 15. That's as long as I'd really like us to be waiting here as this will be holding up the serve loop when it's happening. I think 25 seconds is also long enough that we are justified marking the upgrade as failed.

Also added an info level log statement saying we're waiting to retry fail over so anyone looking at the logs will know why things aren't progressing for that 25 seconds.

**Comment by adk3798** [comment]
jenkins test make check

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2022-04-15_05:56:09-orch:cephadm-wip-adk-testing-2022-04-14-1844-distro-basic-smithi/

1 infra failure reimaging machines
